sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,902 @@
+a kernel to be enqueued via a `kernel` object.
+The application must enqueue the free function kernel according to its type.
+For example, a free function kernel defined via `SYCL_EXT_ONEAPI_RANGE_KERNEL`
+can be enqueued by calling the `handler::parallel_for` overload taking a
+`range`.
+A free function kernel defined via `SYCL_EXT_ONEAPI_ND_RANGE_KERNEL` can be
+enqueued by calling the `handler::parallel_for` overload taking an `nd_range`.
+A free function kernel defined via `SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL` can be
+enqueued by calling `handler::single_task`.
+
+Attempting to enqueue a free function kernel using a mechanism that does not
+match its type results in undefined behavior.
+Attempting to enqueue a free function kernel with a `range` or `nd_range` whose
+dimensionality does not match the free function kernel definition results in
+undefined behavior.
+
+The application is also responsible for setting the values of any kernel
+arguments when the kernel is enqueued.
+For example, when enqueuing a kernel with `handler::parallel_for` or
+`handler::single_task`, the kernel argument values must be set via
+`handler::set_arg` or `handler::set_args`.
+Failing to set the value of a kernel argument results in undefined behavior.
+The type of the value passed to `handler::set_arg` or `handler::set_args` must
+be the same as the type of the corresponding formal parameter in the free
+function kernel.
+Passing a value with a mismatched type results in undefined behavior.
+
+=== Obtaining the iteration id for a kernel
+
+In a standard SYCL kernel, the iteration ID is passed as a parameter to the
+kernel's callable object.
+However, this is not the case for a free function kernel because the function
+parameters are used to pass the kernel arguments instead.
+Therefore, a free function kernel must obtain the iteration ID in some other
+way.
+Typically, a free function kernel uses the functions specified in
+link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
+sycl_ext_oneapi_free_function_queries] for this purpose.
+
+=== Address space of kernel arguments
+
+The arguments to a free function kernel are in the private address space.
+As a result, a kernel can modify its arguments, but the modification is visible
+only within the work-item.
+
+[_Note:_ This applies only to the arguments themselves, not to memory that the
+arguments point to.
+For example, with a USM pointer argument, the pointer argument itself is in the
+private address space, but the memory it points to is in the global address
+space.
+_{endnote}_]
+
+=== Interaction with kernel properties
+
+If the implementation supports
+link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
+sycl_ext_oneapi_kernel_properties], a free function kernel may be decorated
+with these properties by applying the properties to the function declaration as
+illustrated below.
+
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::work_group_size<32>)
```

Comment:
I still find this quite dissatisfying.  I'm worried that if the `FUNCTION_PROPERTY` mechanism doesn't work for this case, it isn't actually going to work for any of the other cases. #10540 is pointing to sycl_ext_oneapi_free_function_kernels as a precedent to justify introducing yet another macro instead of using the property mechanism.

Could we add an issue saying something about how we'd ideally like to unify the mechanism used to declare all of these special behaviors and how they apply to functions?

Relevant discussion: https://github.com/intel/llvm/pull/10540#discussion_r1417519332

Reply:
I added an issue in 0fe104f983c9602f638681c2e08b377f99470676.

However, I'd like to come to a final decision on this soon because it will affect all code using this extension.

Reply:
We decided to go with the unified behavior.  We no longer need the special macros like `SYCL_EXT_ONEAPI_RANGE_KERNEL`, and we can use compile-time properties instead.  I made this change in aa94401c618610c6c8e375ec94540ec4a1951aa8.